statsmodels/regression/heckman.py

```diff
@@ -0,0 +1,663 @@
+        sigma = np.exp(log_sigma)
+        rho = np.tanh(atanh_rho)
+
+        # line the data vectors up
+        Z_zbeta_aligned = Z.dot(zbeta)
+
+        X_xbeta = X.dot(xbeta)
+        X_xbeta_aligned = np.empty(self.nobs_total)
+        X_xbeta_aligned[:] = np.nan
+        X_xbeta_aligned[D] = X_xbeta
+        del X_xbeta
+
+        Y_aligned = np.empty(self.nobs_total)
+        Y_aligned[:] = np.nan
+        Y_aligned[D] = Y
+
+        # create an array where each row is the log likelihood for the corresponding observation
+        norm_cdf_input = (Z_zbeta_aligned+(Y_aligned-X_xbeta_aligned)*rho/sigma) / np.sqrt(1-rho**2)
+        norm_cdf_input[~D] = 0  # dummy value
+
+        ll_obs_observed = \
+            np.multiply(D,
+                norm.logcdf(norm_cdf_input) - \
+                (1./2.)*((Y_aligned-X_xbeta_aligned)/sigma)**2 - \
+                np.log(np.sqrt(2*np.pi)*sigma))
+        ll_obs_observed[~D] = 0
+        ll_obs_notobserved = \
+            np.multiply(1-D,
+                norm.logcdf(-Z_zbeta_aligned))
+
+        ll_obs = ll_obs_observed + ll_obs_notobserved
+
+        # return log likelihood by observation vector
+        return ll_obs
+
+    def score(self, params):
+        '''
+        Gradient of log-likelihood evaluated at params
+        '''
+        #this is the numerical approx func taken from
+        # base.model.GenericLikelihoodModel
+
+        kwds = {}
+        kwds.setdefault('centered', True)
+        return approx_fprime(params, self.loglike, **kwds).ravel()
+
+    def jac(self, params, **kwds):
+        '''
+        Jacobian/Gradient of log-likelihood evaluated at params for each
+        observation.
+        '''
+        #this is the numerical approx func taken from
+        # base.model.GenericLikelihoodModel
+
+        #kwds.setdefault('epsilon', 1e-4)
+        kwds.setdefault('centered', True)
+        return approx_fprime(params, self.loglikeobs, **kwds)
+
+    def hessian(self, params):
+        '''
+        Hessian of log-likelihood evaluated at params
+        '''
+        #this is the numerical approx func taken from
+        # base.model.GenericLikelihoodModel
+
+        from statsmodels.tools.numdiff import approx_hess
+        # need options for hess (epsilon)
+        return approx_hess(params, self.loglike)
+
+
+
+    def predict(self, params, exog=None):
+        """
+        Return linear predicted values from a design matrix.
+
+        Parameters
+        ----------
+        exog : array-like
+            Design / exogenous data
+        params : array-like, optional after fit has been called
+            Parameters of a linear model
+
+        Returns
+        -------
+        An array of fitted values
+
+        Notes
+        -----
+        If the model has not yet been fit, params is not optional.
+        """
+        if exog is None:
+            exog = self.exog
+        return np.dot(exog, params)
+
+        if self._results is None and params is None:
```

Comment:
parts here are not reached. can be deleted
I guess left over from an old pattern attaching `_results` to model